### ISSUE TITLE
Remove all UIWebView code?

### ISSUE THREAD

**Issue by jcesarmobile** [apache/cordova-ios#661]
People has started receiving this warning when submitting apps that have references to UIWebView. 

> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs.

Doesn't matter if they are using just WKWebView, as long as Apple sees "UIWebView" in the app code, they show the warning, and cordova-ios is full of "UIWebView" references. (Even react native, flutter and Capacitor users have received the warning)

**There is currently no solution to this problem for Cordova apps.**

So, instead of integrating WKWebView and making it possible to switch from UIWebView to WKWebView on runtime as it was planned, maybe we should just remove all UIWebView code on next major release?

~~There is no deadline, I've contacted an Apple developer that has been tweeting about moving to WKWebView to see if he can provide more information, will update if I get a response.~~

Apple finally announced the [deadline](https://developer.apple.com/news/?id=12232019b):

> The App Store will no longer accept new apps using UIWebView as of April 2020 and app updates using UIWebView as of December 2020.

**Comment by fjaeger** [comment]
I am not sure whether it is necessary to remove all the references to `UIWebView`. There is even some `UIWebView` code being executed when `WKWebView` is configured as the main engine. I know at least of `CDVUserAgentUtil` that is still using `UIWebView` – independent of the engine configuration.

**Comment by eun-choi** [comment]
It's vert important issue in the near future, because UIWebView been deprecated on iOS 13.
That means we would be rejected by App Store Review at some point in the near future.

https://developer.apple.com/documentation/uikit/uiwebview

**Comment by janpio** [comment]
> It's very important issue in the near future, because UIWebView been deprecated on iOS 13.
> That means we would be rejected by App Store Review at some point in the near future.

That is fortunately not true. Things can be deprecated for a long time before they are removed or their usage affect App Store review.

Which of course does not mean that the previously posted thoughts are wrong in any way - it just doesn't mean this has to lead to problems in the near future. It just definitely _could_.

**Comment by nielsjakob** [comment]
> > It's very important issue in the near future, because UIWebView been deprecated on iOS 13.
> > That means we would be rejected by App Store Review at some point in the near future.
> 
> That is fortunately not true. Things can be deprecated for a long time before they are removed or their usage affect App Store review.
> 
> Which of course does not mean that the previously posted thoughts are wrong in any way - it just doesn't mean this has to lead to problems in the near future. It just definitely _could_.

UIWebView has been deprecated for well over a year though, so I fear the "long time" is coming to an end very soon. So I'd second that this is a rather urgent issue.

Here's a Twitter thread from one Apple dev on the topic:
https://twitter.com/NiklasMaerz/status/1167031874738434048

**Comment by dpogue** [comment]
Just to set expectations: There's probably about 1 week of full-time work to be done by somebody familiar with the cordova-ios and WKWebView plugin codebases to resolve this issue.

There is currently nobody working full-time on Cordova. We are all volunteers with limited time to contribute.

**Comment by derman10** [comment]
Same email from Apple to uploading my app to TestFight.

Im building iOS app from ionic 3/cordova and not use UIWebView, just WebView on my config.xml.

`<plugin name="cordova-plugin-ionic-webview" spec="^4.0.0" />`

Any Solution??

**Comment by jcesarmobile** [comment]
Apple warns because cordova-ios has the UIWebView code, doesn't matter if you use it or not, the code is there.

But it's a warning, we don't know for sure how long will take until Apple starts rejecting apps because of that.

Once they start rejecting apps, the only solution will be to update your cordova-ios version and all plugins that have references to UIWebView to newer versions that don't have those references (they don't exist yet, that's what this issue is for).

**Comment by fjaeger** [comment]
I created a compile-time decision for WKWebView over UIWebView in Pull Request #662

**Comment by spinninghamster** [comment]
iOS 13 will be released around 17 September 2019. Could be that Apple will start rejecting apps after that date..

**Comment by dpogue** [comment]
Apple is not going to start rejecting apps that use UIWebView with less than a month of notice. For the moment, it is safe to ignore the warning.

No version of cordova-ios uses WKWebView by default.

**Comment by spinninghamster** [comment]
> Apple is not going to start rejecting apps that use UIWebView with less than a month of notice. For the moment, it is safe to ignore the warning.
> 
> No version of cordova-ios uses WKWebView by default.

That's somewhat good news. 

I was referring to this: https://twitter.com/NiklasMaerz/status/1167031874738434048 where Niklas wrote:

> WKWebView is a plugin and UIWebView is part of the core platform. This means we only use WKWebView but there are still unused references of UIWebView in code.

I thought he was speaking about Cordova; but maybe he was speaking about a specific own app.

**Comment by NiklasMerz** [comment]
Just to clarify. I was speaking about our own app with Cordova. We are using the WKwebview plugin and InAppBrowser.

I guess many app developers switched to WKwebview by now.

**Comment by fabuhelow** [comment]
> Same email from Apple to uploading my app to TestFight.
> 
> Im building iOS app from ionic 3/cordova and not use UIWebView, just WebView on my config.xml.
> 
> `<plugin name="cordova-plugin-ionic-webview" spec="^4.0.0" />`
> 
> Any Solution??

Please follow the following link which will help you to move from UIwebview to WKwebview

https://ionicframework.com/docs/v3/wkwebview/

**Comment by spinninghamster** [comment]
When cordova-ios migrates from UIWebView to WKWebView, how will LocalStorage in apps be handled? Will there be a way to migrate them over?

**Comment by timbru31** [comment]
That's a good question @spinninghamster! One could argue that a breaking change means no transition of user data, since e.g. plugins for LocalStorage and IndexedDB already exist (e.g., https://github.com/maklesoft/cordova-plugin-migrate-localstorage & https://github.com/pointmanhq/cordova-plugin-ionic-migrate-storage

**Comment by EdwinRozario** [comment]
@fabuhelow Did that solution work for you? My project is running on Ionic 5.1.0. By default its should be using `WKWebView`. Still receiving the warning. I have tried upgrading to the latest version of `cordova-plugin-ionic-webview`, which is 4.1.1 with no success.

I dont think this issue is due to the app using UIWebview. Any reference of it in any Cordova plugins are causing these warnings. I hope we will be resolved or someone will give us an idea of when it will be resolved.

**Comment by guigons** [comment]
Same Here!! 

ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs . See https://developer.apple.com/documentation/uikit/uiwebview for more information.

I think it's a big problem if apple starts to block submissions from IONIC.

=( =( =( =( =( =( =( =( =( =( =( =( =( =( =( =(  ... worried!

**Comment by sts-ryan-holton** [comment]
I'd strongly consider prepping some time for this issue, what we don't want is it to suddenly become depreciated upon a later iOS 13 release and then the apps suddenly can't be built, if it's true that it'll take the best part of a week to fix, let's ensure backwards compatibility with older versions of Cordova and ensure that it can truly be fixed promptly when it happens.

**Comment by bpresles** [comment]
@sts-ryan-holton I've worked already on fixing this potential issue. See PR #663.

**Comment by EdwinRozario** [comment]
Thanks @bpresles . For us working with Ionic, we will to wait for the Ionic team to upgrade `cordova-ios` in `cordova-plugin-ionic-webview` before we can start testing this.

**Comment by janpio** [comment]
> For us working with Ionic, we will to wait for the Ionic team to upgrade `cordova-ios` in `cordova-plugin-ionic-webview` before we can start testing this.

This doesn't make sense. You install `cordova-ios` into your Cordova app that uses Ionic, but the Cordova iOS version is not set by Ionic and definitely not by `cordova-plugin-ionic-webview`.

**Comment by EdwinRozario** [comment]
Yes. But `cordova-plugin-ionic-webview` has a `cordova-ios` dependency. I thought the plan was to get rid of any reference of `UIWebView` from project. If simply updating `cordova-ios` will fix the issue, i will be more than happy.

**Comment by janpio** [comment]
> But `cordova-plugin-ionic-webview` has a `cordova-ios` dependency

Not that I am aware of. It depends on `cordova-ios`, and also requires a specific min version, but that's about it.

**Comment by jcesarmobile** [comment]
It doesn’t have any dependency to cordova-ios, it just requires cordova-ios >= 4.0.0 installed, so once it’s fixed in cordova-ios you update to latest version, and that’s all.

**Comment by bpresles** [comment]
@EdwinRozario It's you who decide which version of cordova-ios you want to use with your Ionic app.
In fact I've used my PR with a Ionic app (Ionic 1.x, not even 2, 3 or 4 ) ;) Also the `cordova-plugin-ionic-webview` plugin will still work as the plugin mechanism of engine is kept, so if you prefer to use the engine from this plugin instead of the embedded CDVWKWebViewEngine in this PR, feel free to install and use it ;)

But indeed as `cordova-plugin-ionic-webview` use similar class names than the one of CDVWKWebViewEngine from `cordova-plugin-wkwebview-engine` I embedded in this PR, it'll require an update.
I already done a fork with renamed classes (using ION instead of CDV on all classes of the plugin): https://github.com/bpresles/cordova-plugin-ionic-webview.git with the associated PR: https://github.com/ionic-team/cordova-plugin-ionic-webview/pull/430

**Comment by desmeit** [comment]
@bpresles 

i tried your 4.1.2 version. I get this Error:

`.../Plugins/cordova-plugin-ionic-webview/IONAssetHandler.m:3:9: 'CDVWKWebViewEngine.h' file not found`

**Comment by bpresles** [comment]
@desmeit Sorry, the tag is not correct (not including the latest commit). I’ve re-tagged it correctly (make sure you remove all previously fetched version).

**Comment by bpresles** [comment]
@desmeit At the end after reading the https://github.com/apache/cordova-discuss/pull/110 discussion I decided to rename classes in the embedded WKWebView engine of this PR, so that all existing engine plugin work out of the box without any change.

So if you use the latest version of my fork, than you can use any existing engine plugin you want without modification.

**Comment by fjaeger** [comment]
I don't have a strong preference between #662 and #663 as long as the underlying issue is resolved eventually. I agree with @bpresles that removing the UIWebView code is the logical step as there would be no gain in keeping the legacy code when we have a working WKWebView-based implementation.
Those that still require UIWebView (meaning being incompatible with the AppStore at some point) could still use a previous version of this project.

**Comment by bpresles** [comment]
@fjaeger For information, I also created an engine plugin based on current UIWebView engine so that even if someone want to use UIWebView with future cordova-ios, they can: https://github.com/bpresles/cordova-plugin-uiwebview-engine

I took care to rename the classes so that it doesn't conflict with any existing one, which makes the plugin usable even with versions of `cordova-ios` that have UIWebView engine as default (which is meaningless of course ;) ))

**Comment by BrainstemStudio** [comment]
I recently uploaded an app to app store and was approved for public beta testing. I do use ionic's WKWebView but did receive a warning about UIWebView. So at least open beta gets approved despite the warning.

Just adding to the conversation.

**Comment by spinninghamster** [comment]
> I recently uploaded an app to app store and was approved for public beta testing. I do use ionic's WKWebView but did receive a warning about UIWebView. So at least open beta gets approved despite the warning.
> 
> Just adding to the conversation.

Question is: Will repeated uploads (new versions) with that error message without taking action (removing UIWebView) lead to app store rejections?

**Comment by jcesarmobile** [comment]
> Question is: Will repeated uploads (new versions) with that error message without taking action (removing UIWebView) lead to app store rejections? I am not suggesting. I just don't know so I am wondering if anyone does.

No, it won't matter how many warnings you got, there will be a day (unknown for now) where any app submitted with UIWebView will be rejected, whether it was warned multiple times or none. That day will be the same for everybody, we are still trying to get an official date from Apple.

**Comment by Vito96** [comment]
Sorry but I don't understand, at this moment there are any solutions for this problem? (I currently use the WkWebview but i receive anyway this warning)

**Comment by derman10** [comment]
question is: Cordova team is working on a solution???
And what will we do? Will we have to migrate our projects to Capacitor?

**Comment by breautek** [comment]
> Sorry but I don't understand, at this moment there are any solutions for this problem? (I currently use the WkWebview but i receive anyway this warning)

No solution currently. You get the warning because even if you're using the `WkWebview`, `UIWebView` class still referenced in the core cordova ios libraries.

**Comment by bpresles** [comment]
@derman10 @Vito96 
See PR #662 or #663 . You can use one of these fork if you want to get rid of warning message today, but it's just a warning for now, so your app(s) won't be rejected even with current official cordova-ios.

**Comment by breautek** [comment]
> question is: Cordova team is working on a solution???
> And what will we do? Will we have to migrate our projects to Capacitor?

Cordova is working on it, but keep in mind that they are a team of volunteers. You can keep track of the discussion on how exactly Cordova will solve this problem at https://github.com/apache/cordova-discuss/pull/110

**Comment by spinninghamster** [comment]
> #663

It's great to know an emergency solution is available if needed but me personally I will wait for the official release. I'd like to take this moment to thank the team of volunteers which keeps Cordova going. Everyone complains fast when things break but few people take a second to appreciate the hours of effort put into it. So **thank you** and if there's anything the community can do to help don't be afraid to ask.

**Comment by vtellez** [comment]
> > #663
> 
> It's great to know an emergency solution is available if needed but me personally I will wait for the official release. I'd like to take this moment to thank the team of volunteers which keeps Cordova going. Everyone complains fast when things break but few people take a second to appreciate the hours of effort put into it. So **thank you** and if there's anything the community can do to help don't be afraid to ask.

+1

**Comment by XELoh** [comment]
Hi, I just face this issue, I check my plugin have installed inAppBrowser, and I am using ionic 3, got any solution can fix this?

**Comment by NiklasMerz** [comment]
Please read the whole thread and this discussion. https://github.com/apache/cordova-discuss/pull/110

A solution is being discussed. Please be patient everybody who gets the warning.

**Comment by pauldrummond** [comment]
> > It's very important issue in the near future, because UIWebView been deprecated on iOS 13.
> > That means we would be rejected by App Store Review at some point in the near future.

I just had feedback from Apple on a Cordova-based app submitted for a client. While the app was rejected for content issues, nothing was raised regarding UIWebView. Looks like the code is still being accepted.

**Comment by manish2535** [comment]
> 
> There is no deadline, I've contacted an Apple developer that has been tweeting about moving to WKWebView to see if he can provide more information, will update if I get a response.

Did you ever hear back from Apple about any Deadlines for this deprecation? Thanks.

**Comment by GulinSS** [comment]
At my experience they begun treat it as hard fail.

**Comment by jcesarmobile** [comment]
> > There is no deadline, I've contacted an Apple developer that has been tweeting about moving to WKWebView to see if he can provide more information, will update if I get a response.
> 
> @jcesarmobile Did you ever hear back from Apple about any Deadlines for this deprecation? Thanks.

No, they didn’t give any date, if not I would have updated.


> At my experience they begun treat it as hard fail. But they forgot to change text in email message. Still say about "will stop".

Unless you mean your app got rejected and you are the first person to report it, we have not heard anybody whose app has been rejected because of this.

**Comment by nielsjakob** [comment]
> At my experience they begun treat it as hard fail. But they forgot to change text in email message. Still say about "will stop".

@GulinSS Could you please clarify - did your App get recently rejected **because** of the UIWebView issue?

**Comment by nielsjakob** [comment]
> @GulinSS Could you please clarify - did your App get recently rejected **because** of the UIWebView issue?

Okay, I can offer the count-evidence: my app was just approved, even though i got the UIWebView warning :)

**Comment by captaincaius** [comment]
I can confirm that it's still just a "warning" (I got to "Pending Developer Release" even with that email).

**Comment by thiagomatrix** [comment]
Hi, in blog ionicframework: Recently developers were sent out an email from Apple about Web Based Apps (noted as HTML5 Apps) in the App store. Before we jump into the post, let me just say that **this update does not affect Ionic apps in any way.** https://ionicframework.com/blog/making-sense-of-apples-tos/

**Comment by markerio** [comment]
@thiagomatrix This actually does not affect Cordova-based apps in general.

@captaincaius @nielsjakob Thanks for the update.

**Comment by foolishsailor** [comment]
In the move to UIWebView there is an issue with XMLHttpRequest not working.

There is a current fix for this by Oracle:  [CB-10143](https://issues.apache.org/jira/browse/CB-10143) - are there plans to bake this into Cordova-Ios once the switch is made?

**Comment by markerio** [comment]
@foolishsailor What do you mean by:
> In the move to UIWebView there is an issue with XMLHttpRequest not working.

This thread is created to discuss a certain issue regarding the old UIWebView. Apple will stop accepting submissions of apps that use UIWebView APIs or even any references to it. This means you need to switch to WKWebView.

And I confirm that XMLHttpRequest is working well with WKWebView with no issues, tested on iOS 12.x or even the new iOS 13 version.

I'm here trying to help, what javascript library are you using?

**Comment by foolishsailor** [comment]
Oops - sorry mis-typed.  Meant in the move to **WKWebView**

I have not made the change to WKWebView but am beginning to prepare to do so.

Our app however has a significant number of plugins including the use of raw sockets found in this [plugin](https://github.com/KoenLav/cordova-plugin-chrome-apps-sockets-tcp) and we are also worried about how many plugins are potentially still using UIWebView as well...

Really enjoy cordova and comfortable with it but wondering if time to move to react native?

**Comment by dpogue** [comment]
**If you are having issues with the cordova-plugin-wkwebview-engine, please report them [on that repository](https://github.com/apache/cordova-plugin-wkwebview-engine/issues).**

This issue is **not** the place for those discussions, this is intended to plan out the migration strategy for future versions of cordova-ios.

**Comment by foolishsailor** [comment]
Not looking to discuss specific issues – however understanding how bugs in key functionality are or aren’t going to be dealt with in the migration are important for everyone to know

**Comment by GulinSS** [comment]
I think they detect somehow is your app Cordova based or it uses Cordova in existing application. 🤔

**Comment by bpresles** [comment]
> I think they detect somehow is your app Cordova based or it uses Cordova as an additional library to existing application. 🤔

No, they detect the use of UIWebView APIs. Building a Cordova app with one of the forked version of `cordova-ios` mentionned here remove the warning. I've already used my fork (#663) in two Ionic 1.x apps and I don't get the warning got either of them with my fork.

**Comment by GulinSS** [comment]
> No, they detect the use of UIWebView APIs. Building a Cordova app with one of the forked version of `cordova-ios` mentionned here remove the warning. I've already used my fork (#663) in two Ionic 1.x apps and I don't get the warning got either of them with my fork.

By that you are just approving my initial sentence there: https://github.com/apache/cordova-ios/issues/661#issuecomment-533491490

**Comment by breautek** [comment]
> (if you can not deploy your app without erasing UIWebView components)

A warning is not an error. As previously stated several times, you can still publish your app, even if you are currently reference the `UIWebview` class in your app. The warning serves as an official notice that changes will be required in the future. There is no official deadline yet when Apple will start to reject apps for using `UIWebview`.

**Comment by GulinSS** [comment]
For some reason our app was unable to be passed this "warning" but we had no other warnings. When we applied https://github.com/apache/cordova-ios/pull/663 next attempt was successful.

**Comment by markerio** [comment]
> For some reason our app was unable to be passed this "warning"

@GulinSS What do you mean by that?

I may not agree with that. I can confirm that this is just a warning message sent by Apple. Please read @breautek comment above. Receiving this message does NOT mean that your app will be rejected or even prevented from being uploaded to AppleConnect.

**Comment by chinzyr** [comment]
Hi Everyone, Just wanted to let you guys know that we submitted our Cordova based app last friday and it was approved by the app store.

**Comment by XELoh** [comment]
> Hi Everyone, Just wanted to let you guys know that we submitted our Cordova based app last friday and it was approved by the app store.

without any warning message?

**Comment by chinzyr** [comment]
> > Hi Everyone, Just wanted to let you guys know that we submitted our Cordova based app last friday and it was approved by the app store.
> 
> without any warning message?

@XELoh  Yes, I got warnings when I uploaded the app to test flight, but none for the submission.

**Comment by gnrmarcel** [comment]
my app for testflight was not accepted

> Though you are not required to fix the following issues, we wanted to make you aware of them:
> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs . See https://developer.apple.com/documentation/uikit/uiwebview for more information. 
> Best regards,
> The App Store Team

**Comment by derman10** [comment]
> my app for testflight was not accepted
> 
> > Though you are not required to fix the following issues, we wanted to make you aware of them:
> > ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs . See https://developer.apple.com/documentation/uikit/uiwebview for more information.
> > Best regards,
> > The App Store Team

Still this is a warning, not reject.

**Comment by breautek** [comment]
>my app for testflight was not accepted

You're app was not rejected because of `UIWebView` deprecation.

It says so in the first sentence...

>Though you are not required to fix the following issues, we wanted to make you aware of them:

**Comment by gnrmarcel** [comment]
sorry, my mistake it was because of ITMS-90683: Missing Purpose String in Info.plist - but thats not the correct thread for this

ITMS-90683: Missing Purpose String in Info.plist - Your app's code references one or more APIs that access sensitive user data. The app's Info.plist file should contain a NSBluetoothAlwaysUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data. Starting Spring 2019, all apps submitted to the App Store that access user data are required to include a purpose string. If you're using external libraries or SDKs, they may reference APIs that require a purpose string. While your app might not use these APIs, a purpose string is still required. You can contact the developer of the library or SDK and request they release a version of their code that doesn't contain the APIs. Learn more (https://developer.apple.com/documentation/uikit/core_app/protecting_the_user_s_privacy).

**Comment by suraneti** [comment]
I've received this warning email from Apple today, my app still appears on TestFlight and can download on devices.

**Comment by Kraudi** [comment]
I also got the mail from Apple when I was uploading my build to itunesconnect. Though my app was quickly approved and was released with no problems.

**Comment by spinninghamster** [comment]
I submitted an update **yesterday** and did **NOT** get any warning message from Apple. The app is currently "In Review" and still no warning anywhere. Neither during uploading nor an email. Has anyone else experienced this?

**Comment by drawcard** [comment]
My experience has been that I got the warning emails when publishing to Testflight but once the app was live in the App Store, no warnings were sent.

**Comment by markerio** [comment]
@spinninghamster I'm glad to share my experience with you.

Last week we uploaded a new build, we haven't received any warning emails or even alerts through appstoreconnect's console, everything went so smooth. On the same day the app got approved and released to the store.

We use the master release of `cordova-plugin-wkwebview-engine`, not the forked version (as suggested by other folks). Also, it's worth mentioning that we use `cordova-plugin-browsertab-themeable` NOT `cordova-plugin-inappbrowser`.

Wish you the best of luck.

**Comment by spinninghamster** [comment]
I also use the original `cordova-plugin-wkwebview-engine` but strangely I **do** use `cordova-plugin-inappbrowser` but I am not receiving the error..

**Comment by markerio** [comment]
> I also use the original cordova-plugin-wkwebview-engine but strangely I do use cordova-plugin-inappbrowser but I am not receiving the error..

@spinninghamster That's a great piece of information, thanks for sharing. I can't find an answer for this, however what happened to you happened to us. As the status of you app is "In Review", just Sit back and Relax. Everything will be okay!

**Comment by XELoh** [comment]
> I also use the original `cordova-plugin-wkwebview-engine` but strangely I **do** use `cordova-plugin-inappbrowser` but I am not receiving the error..

I also same with you, but I get the ITMS-90809 warning at testFlight only, is everyone still like this?

**Comment by desmeit** [comment]
I'm still not able to solve the problem yet. With other Cordova Apps I don't get the issue. I compare them and don't know why i get the error with one app and not with the other.

Is there a way I can search the plugins to find out where the bug might be?

**Comment by suraneti** [comment]
@desmeit it's not a bug, it just a warning from Apple about you're using UIWebView in your code.

**Comment by longhai1221** [comment]
Apple is still warning me about this issue :(. Should I try to submit?

**Comment by dmac81** [comment]
@longhai1221 its just a warning, you can submit without changes until it becomes an error at some unknown day in the future, which would likely be later next year (guessing)

**Comment by derman10** [comment]
Why are Cordova Developers waiting for Apple not to let applications publish to provide the real solution to this problem? it's been 3 months!

**Comment by breautek** [comment]
> Why Cordova Developers are waiting for Apple not to let applications publish to provide the real solution to this problem? it's been 3 months!
> One day the warning will stop being that to become a stop.

[This](https://github.com/apache/cordova-discuss/pull/110) is why. This is not as simple as flipping a switch. Also keep in mind that Cordova is entirely community and volunteer driven.

**Comment by longhai1221** [comment]
> @longhai1221 its just a warning, you can submit without changes until it becomes an error at some unknown day in the future, which would likely be later next year (guessing)

thank you! I will try. aww. the first time I get this =..=

**Comment by yiwong2001** [comment]
Dear Developer,

We identified one or more issues with a recent delivery for your app, "XXXX" 1.0.1 (1.0.1). Your delivery was successful, but you may wish to correct the following issues in your next delivery:

ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs . See https://developer.apple.com/documentation/uikit/uiwebview for more information.

After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to App Store Connect.

Best regards,

**Comment by markerio** [comment]
@yiwong2001 please read the message carefully, it clearly says:
> **Your delivery was successful**, but you **may wish** to correct the following issues in your next delivery..

So, as other colleges have said, it’s just a warning message. You’re welcome to skip the message if you like. Wish you luck.

**Comment by erikyuzwa** [comment]
Trying to navigate towards a solution myself, but my Cordova-based app went through submission without issue yesterday to Apple - and yes I still received the deprecation warning (which was the catalyst that led me to this thread).

**Comment by breautek** [comment]
> Trying to navigate towards a solution myself, but my Cordova-based app went through submission without issue yesterday to Apple - and yes I still received the deprecation warning (which was the catalyst that led me to this thread).

https://github.com/apache/cordova-ios/pull/715 has landed in master, but haven't been released yet. It will be released in 5.1.0. I believe a release is being prepared soon, but not 100% sure on that. 

This PR will allow you to disable all `UIWebView` references using a flag. Using a `WkWebVIew` plugin will become mandatory with this flag enabled.

If you want to try this, it should be available at `cordova-ios@nightly`

**Comment by RichardBoyder** [comment]
Hi guys. A Strange Question I know in this forum.. But Are most people moving over to capacitor?  Is it Safe to say that the developers who were focusing more on Cordova has taken a slightly different focus? OR Are we Still very serious about Cordova and its future?   (Given Apple issues with soon to be deprecated UiWebView.)

This isn't to blame the developers at all , Cordova is brilliant, Its just a simple question Since I now have to make a decision to leave Cordova and go with Capacitor that Ionic has introduced Since Apple Store issue.  Business is pushing so its either or ... And I hate to leave Cordova since it has sooooo many awesome plugins and I have built my custom plugin for it which is working wonders.

**Comment by suraneti** [comment]
@RichardBoyder I not sure about Capacitor but for Cordova right now is driven by the community.

**Comment by RichardBoyder** [comment]
> @RichardBoyder I'm not sure about Capacitor but for Cordova right now is driven by the community.

Understood. Thanks for the reply.

**Comment by EiyuuZack** [comment]
> @RichardBoyder I'm not sure about Capacitor but for Cordova right now is driven by the community.

@suraneti what exactly do you mean by this? Does Apache no longer have anything to do with Cordova?

**Comment by NiklasMerz** [comment]
> > @RichardBoyder I'm not sure about Capacitor but for Cordova right now is driven by the community.
> 
> @suraneti what exactly do you mean by this? Does Apache no longer have anything to do with Cordova?

I don't think this is the right place to discuss the Cordova vs Capacitor topic, since this is about an important Cordova issue to be discussed.

I would like to add a short personal note about that nevertheless:

Cordova still an active Apache project worth investing in. It's maintained by volunteers, but probably @suraneti is meaning that nobody is paid to work full-time on Cordova. This means (to me) you may need to be more patient and open to fixing issues and contributing yourself.

**Comment by RichardBoyder** [comment]
All understood. We seriously need a fix for Cordova so we don't have issues when apple decides to cancel our binary uploads. So a very serious matter.

Anyone who can comment and perhaps We can make a path to get it sorted so we don't have to worry about our Cordova apps being rejected in future.. its kind of unnecessary stress on us developers now and I really hope anyone who has Power here to fix this for us will look seriously at this issue. Please need a answer. Will we be able to rest in the knowledge that this will be sorted within next 2 months or so...  ? I'm asking the Cordova Devs here who can do it and who can push a new major version of Cordova so we can all implement this?

**Comment by RichardBoyder** [comment]
> > > @RichardBoyder I'm not sure about Capacitor but for Cordova right now is driven by the community.
> > 
> > 
> > @suraneti what exactly do you mean by this? Does Apache no longer have anything to do with Cordova?
> 
> I don't think this is the right place to discuss the Cordova vs Capacitor topic, since this is about an important Cordova issue to be discussed.
> 
> I would like to add a short personal note about that nevertheless:
> 
> Cordova still an active Apache project worth investing in. It's maintained by volunteers, but probably @suraneti is meaning that nobody is paid to work full-time on Cordova. This means (to me) you may need to be more patient and open to fixing issues and contributing yourself.

I wish I had the skills to fix this haha (I don't) ... otherwise I would be doing it right now... So many hundreds of people are relying on Cordova for their apps on store and I'm sure everyone is worried.

**Comment by spinninghamster** [comment]
> All understood. We seriously need a fix for Cordova so we don't have issues when apple decides to cancel our binary uploads. So a very serious matter.
> 
> Anyone who can comment and perhaps We can make a path to get it sorted so we don't have to worry about our Cordova apps being rejected in future.. its kind of unnecessary stress on us developers now and I really hope anyone who has Power here to fix this for us will look seriously at this issue. Please need a answer. Will we be able to rest in the knowledge that this will be sorted within next 2 months or so... ? I'm asking the Cordova Devs here who can do it and who can push a new major version of Cordova so we can all implement this?

It has already been fixed. Read:
https://cordova.apache.org/announcements/2019/11/25/cordova-ios-release-5.1.0.html

**Comment by RichardBoyder** [comment]
Unfortunately not. Moved to 5.1.0 ages ago when it came out and still get the apple warning.  did everything recommended there and its still happening when push to store.

**Comment by breautek** [comment]
> Unfortunately not. Moved to 5.1.0 ages ago when it came out and still get the apple warning. did everything recommended there and its still happening when push to store.

Then you likely have a plugin that still references `UIWebview` api. A common one that I can think of is the inappbrowser plugin. So I would inspect your plugins.

**Comment by RichardBoyder** [comment]
> > Unfortunately not. Moved to 5.1.0 ages ago when it came out and still get the apple warning. did everything recommended there and its still happening when push to store.
> 
> Then you likely have a plugin that still references `UIWebview` api. A common one that I can think of is the inappbrowser plugin. So I would inspect your plugins.

That is very true. Could be the case. I guess the plugin authors haven't updated their plugins yet to sort this out. But could certainly find and comment out I guess in the interim ... if it doesn't break the app.

**Comment by jacobweber** [comment]
OK, the [deadline](https://developer.apple.com/news/?id=12232019b) is upon us:

> The App Store will no longer accept new apps using UIWebView as of April 2020 and app updates using UIWebView as of December 2020.

**Comment by spinninghamster** [comment]
> OK, the [deadline](https://developer.apple.com/news/?id=12232019b) is upon us:
> 
> > The App Store will no longer accept new apps using UIWebView as of April 2020 and app updates using UIWebView as of December 2020.

Now: _"Oh, a year of time ..all good"_

In November 2020: _"Oh my god, we're running out of time! The 10 most popular Cordova plugins are all not updated yet!"_

**Comment by desmeit** [comment]
@breautek Is there a way to search the plugins to find UIWebview api? What is the exact keyword?

**Comment by desmeit** [comment]
In my case the following plugins caused the issue:

facebook4 3.1.0
Googleplus 7.0.1
cordova-plugin-decimal-keyboard

i tested it step by step. delete plugin by plugin one after the other until the message disappears. not nice, but I couldn't find out otherwise.

**Comment by JamesCoyle** [comment]
I've recently created a new Cordova project specifically for iOS and I'm getting this issue. I'm using cordova-ios v5.1.1 and the only plugin I'm using is cordova-plugin-whitelist v1.3.4. What can I do about this?

**Comment by breautek** [comment]
> I've recently created a new Cordova project specifically for iOS and I'm getting this issue. I'm using cordova-ios v5.1.1 and the only plugin I'm using is cordova-plugin-whitelist v1.3.4. What can I do about this?

Add a wkwebview plugin such as [this one](https://github.com/apache/cordova-plugin-wkwebview-engine).

Do pay attention to the required [config.xml](https://github.com/apache/cordova-plugin-wkwebview-engine#configxml) settings.

Do note that plugin is a stopgap, it won't be needed in the next major release of cordova-ios, which will only support wkwebview, and it will be built-in.

**Comment by KevinKelchen** [comment]
I want to give a big THANK YOU to ALL of the contributors (@erisu, @jcesarmobile, etc.) who have helped in any way to add the feature that removes UIWebView references at compile-time.

Just this past week we began receiving emails from Apple about our existing app having UIWebView references even though we were using WKWebView to host our app. We were able to make a small version bump to `cordova-ios` and opt-in with the new flag and it worked beautifully! No more emails from Apple!

FWIW, these are the plugins we have in our project:
```json
"plugins": {
  "cordova-plugin-cartegraph-datepicker": {},
  "cordova-plugin-device": {},
  "cordova-plugin-safariviewcontroller": {},
  "cordova-plugin-whitelist": {},
  "cordova-plugin-screen-orientation": {},
  "cordova-plugin-splashscreen": {},
  "cordova-plugin-statusbar": {},
  "cordova-sqlite-ext": {},
  "cordova-plugin-network-information": {},
  "cordova-plugin-geolocation": {},
  "cordova.plugins.diagnostic": {
    "ANDROID_SUPPORT_VERSION": "26.+"
  },
  "cordova-plugin-camera": {},
  "com-sarriaroman-photoviewer": {},
  "cordova-plugin-image-resizer": {},
  "cordova-plugin-customurlscheme": {
    "URL_SCHEME": "cgone",
    "ANDROID_SCHEME": " ",
    "ANDROID_HOST": " ",
    "ANDROID_PATHPREFIX": "/"
  },
  "cordova-plugin-ionic-keyboard": {},
  "cordova-ios-plugin-no-export-compliance": {},
  "cordova-plugin-ionic-webview": {
    "ANDROID_SUPPORT_ANNOTATIONS_VERSION": "27.+"
  },
  "cordova-plugin-wkwebview-inject-cookie": {},
  "cordova-plugin-cartegraph-cookie-master": {}
}
```

I appreciate the volunteer work of Cordova maintainers and contributors who spend personal time outside of their day jobs to help keep things running smoothly for us! ❤️

**Comment by biswas123** [comment]
@KevinKelchen Does this mean that anyone using cordova-ios: 5.1.1 and usage of latest plugins will be able to pass app submission sucessfully?

**Comment by KevinKelchen** [comment]
@biswas123, using `cordova-ios` 5.1.1 and the new flag will remove UIWebView references automatically for you at compile-time, but it's just those UIWebView references _that are part of `cordova-ios`_.

If you use other plugins that themselves reference UIWebView, removing those UIWebView references has to be addressed separately by those plugins; updating `cordova-ios` and using the new flag will not inherently remove those references.

We did not have any plugins that referenced UIWebView so updating `cordova-ios` was all that was necessary for us to pass app validation successfully.

**Comment by Mathieu-Castets** [comment]
Does anyone know how to check if a build still reference UIWebView like Apple do? 

I have tried to unzip my Appname.ipa file then run `grep -lr "UIWebView" *`. 
It returns a line like `Payload/MyApp.app/MyApp`. Does it mean my build won't pass Apple check?

Even better, how to know which plugins still reference UIWebView?

**Comment by KevinKelchen** [comment]
I don't know enough to comment on analyzing the app binary to know whether it will contain UIWebView usages that Apple won't like.

As far as plugins go, you could search the `plugins` directory of your repo after a `cordova prepare` to see if any plugins use UIWebView.

Otherwise you could try uploading your build to Apple TestFlight.

At least for us, we started receiving an email from Apple with every single upload to TestFlight. After we updated `cordova-ios` and used the new flag those emails completely went away. That was good enough evidence for us.

If you're not using TestFlight or you're not already getting emails from Apple about UIWebView from your uploads to TestFlight, this technique may not work well. But the "proof in the pudding" of whether your app contains any UIWebView-related stuff Apple wouldn't like is to actually run it through Apple's validation and see if it gets flagged (upload to TestFlight or submit the app to the App Store).

**Comment by nulele** [comment]
> I don't know enough to comment on analyzing the app binary to know whether it will contain UIWebView usages that Apple won't like.
> 
> As far as plugins go, you could search the `plugins` directory of your repo after a `cordova prepare` to see if any plugins use UIWebView.
> 
> Otherwise you could try uploading your build to Apple TestFlight.
> 
> At least for us, we started receiving an email from Apple with every single upload to TestFlight. After we updated `cordova-ios` and used the new flag those emails completely went away. That was good enough evidence for us.
> 
> If you're not using TestFlight or you're not already getting emails from Apple about UIWebView from your uploads to TestFlight, this technique may not work well. But the "proof in the pudding" of whether your app contains any UIWebView-related stuff Apple wouldn't like is to actually run it through Apple's validation and see if it gets flagged (upload to TestFlight or submit the app to the App Store).

Same problem.
Sorry @KevinKelchen but I don't understand what "flag" are you talking about... where do I have to opt-in that flag?

Thanks

**Comment by erisu** [comment]
* Use `cordova-ios@5.1.1`
* Add `preference` flag `WKWebViewOnly` to the `config.xml`.

**Must use a `WKWebView` plugin.**
The following example uses the `cordova-plugin-wkwebview-engine` plugin:

* Add plugin `cordova-plugin-wkwebview-engine`, preferably the latest version `1.2.1`
* Add `preference` flag `CordovaWebViewEngine` to the `config.xml`.
* Add `feature` attribute `CDVWKWebViewEngine` to the `config.xml`.

**Example `config.xml` with `cordova-plugin-wkwebview-engine` plugin**
```
<platform name="ios">
    <preference name="WKWebViewOnly" value="true" />

    <feature name="CDVWKWebViewEngine">
        <param name="ios-package" value="CDVWKWebViewEngine" />
    </feature>

    <preference name="CordovaWebViewEngine" value="CDVWKWebViewEngine" />
</platform>
```

**Still getting the warning?**
If you are still getting the warning, most likely a plugin that you are using still references `UIWebView`. You will need to identify which plugin and contact the plugin developers and request them to fix it. They can either remove the references or make use of this flag to wrap the references.

**All official Cordova plugins were updated to use the flag.**

**Using Ionic WKWebView plugin?**
If you are using Ionic's plugin, I believe the `UIWebView` references have already been removed from their latest version. 

* Make sure you are using the latest version.
* Make sure the `WKWebViewOnly` flag is set still for `cordova-ios` platform.

The other `preference` and `feature` arguments seen in the example above are from `cordova-plugin-wkwebview-engine` and may not be needed for Ionic.

If you have any issue or usage questions with the Ionic plugin. Read their docs and request help from their support channels.
